include/net/xia_ether.h

```diff
@@ -240,6 +243,105 @@ static int attach_neigh_addr_to_fib_entry(struct fib_xid_ether_main
 	return 0;
 }
 
+/* Template struct containing hdr manipulation funcs. */
+struct xia_header_ops {
+	/* Creates a new L2 header. */
+	int (*create) (struct sk_buff *skb, struct net_device *dev,
+			   unsigned short type, const void *daddr,
+			   const void *saddr, unsigned int len);
+
+	/* Extracts hrdw address from pkt. */
+	int (*parse)(const struct sk_buff *skb, unsigned char *haddr);
+
+	/* Creates a cached hdr and alocates to a mfxid. */
+	int (*cache)(const struct fib_xid_ether_main *mfxid, struct hh_cache *hh, __be16 type);
+
+	/* Updates the cached hdr on change of address
+	 * type : 1 -> destionation addr
+	 *        0 -> source addr
+	 */
+	void (*cache_update)(struct hh_cache *hh, const struct net_device *dev,
+				const unsigned char *haddr, const int type);
+};
+
+int xia_ether_header_cache(const struct fib_xid_ether_main *mfxid, 
+        struct hh_cache *hh, __be16 type)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	eth = (struct ethhdr *)
+	    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	if (type != htons(ETH_P_XIP))
+		return -1;
+
+	eth->h_proto = type;
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+	memcpy(eth->h_dest, mfxid->neigh_addr->ha, ETH_ALEN);
+	hh->hh_len = ETH_HLEN;
+	return 0;
+}
+
+void xia_ether_header_cache_update(struct hh_cache *hh,
+			     const struct net_device *dev,
+			     const unsigned char *haddr,
+			     const int type)
+{
+	struct ethhdr *eth;
+	eth = (struct ethhdr *)
+		    (((u8 *) hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+	if(type)
+		memcpy(eth->h_dest , haddr, ETH_ALEN);
+	else
+		memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+}
+
+const struct xia_header_ops xia_ether_hdr_ops ____cacheline_aligned = {
+	.create         = eth_header,
+	.parse          = eth_header_parse,
+	.cache          = xia_ether_header_cache,
+	.cache_update   = xia_ether_header_cache_update,
+};
+
+/* Cached Hdr functions. */
+static void mfxid_hh_init(struct fib_xid_ether_main *mfxid)
+{
+	__be16 prot = htons(ETH_P_XIP);
+	struct hh_cache	*hh = &mfxid->cached_hdr;
+
+	write_lock_bh(&mfxid->chdr_lock);
```

Comment:
There are two reasons to remove the `chdr_lock` lock:

1. All code that is not useful now may never be useful, but it will make things harder to maintain and likely harbor bugs. If we were to consider all the cases in which we could add code to XIA that is not being used, but we hoped that could eventually be useful, it would more than double our codebase. Think of how much harder it would've been for you to learn our codebase with twice more code that is not being used. If we eventually find the need for something like the `chdr_lock` lock, we'll add it and we will do it better because we'll have information on how it will be used then.

2. Right now, the `chdr_lock` lock is not even serving its purpose because `mfxid_update_hhs()` changes the header without acquiring the `chdr_lock` lock.